### PR TITLE
Add support for autofilling login text fields

### PR DESCRIPTION
**Pull request by hcsch**
Support autofilling login text fields via Android Autofill with the experimental AndroidX Compose API for it.

Based on https://bryanherbst.com/2021/04/13/compose-autofill/ (with permission).

As I was trying out a prebuilt APK of Element X for Android, I noticed that the login fields could not be autofilled, which seems rather unfortunate since I assume many other users besides myself would like to use the autofill function of their favourite password manager (via Android Autofill).

This implementation of autofill support comes with some minor caveats:
* unlike with traditional Android UI components, the `TextField`s don't have their background changed automatically after a successful autofill (this can be easily added though, if wanted)
* no autofill context menu, unlike with traditional UI components (could probably also be added, if wanted)
* this API is subject to future change, and is marked experimental accordingly (it does however seem to currently be the best way to get autofill working). This being only a temporary API is mentioned here: https://developer.android.com/reference/kotlin/androidx/compose/ui/platform/package-summary#LocalAutofillTree(), where the docs reference `b/138604305` which I presume is some issue tracker reference for "Autofill Semantics" the issue belonging to which I have been unable to find

If I understand correctly, Element is moving towards using SSO portals and OAuth as opposed to plain logins (please correct me if I'm wrong), but before that happens, having autofill for the current UI would also be nice :)

I've tested this code successfully on my real phone with API level 33 (Android 13), didn't test with API level 21 (mentioned in [CONTRIBUTING.md](https://github.com/vector-im/element-x-android/blob/develop/CONTRIBUTING.md#tests)), because it's below the currently configured `minSdk` of 23 (specified in `plugins/src/main/kotlin/Versions.kt`), and failed to test with API level 23, because of some to me seemingly unrelated errors (see below).

<details>
<summary>Click to expand logcat output for first error on AVD with API level 23...</summary>

```logcat
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  FATAL EXCEPTION ElementX Version : 1.0Phone : Android SDK built for x86_64 (6695544 6.0 REL)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  Memory statuses 
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  usedSize   6 MB
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  freeSize   0 MB
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  totalSize   6 MB
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  Thread: main, Exception: java.lang.RuntimeException: Unable to start activity ComponentInfo{io.element.android.x.debug/io.element.android.x.MainActivity}: java.io.FileNotFoundException: /data/user/0/io.element.android.x.debug/databases/session_database.key: open failed: ENOENT (No such file or directory)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2416)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.app.ActivityThread.-wrap11(ActivityThread.java)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.os.Handler.dispatchMessage(Handler.java:102)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.os.Looper.loop(Looper.java:148)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.app.ActivityThread.main(ActivityThread.java:5417)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at java.lang.reflect.Method.invoke(Native Method)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  Caused by: java.io.FileNotFoundException: /data/user/0/io.element.android.x.debug/databases/session_database.key: open failed: ENOENT (No such file or directory)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at libcore.io.IoBridge.open(IoBridge.java:452)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at java.io.FileOutputStream.<init>(FileOutputStream.java:87)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at java.io.FileOutputStream.<init>(FileOutputStream.java:72)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at androidx.security.crypto.EncryptedFile.openFileOutput(EncryptedFile.java:199)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.encrypteddb.passphrase.RandomSecretPassphraseProvider.getPassphrase(RandomSecretPassphraseProvider.kt:47)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.encrypteddb.SqlCipherDriverFactory.create(SqlCipherDriverFactory.kt:39)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.libraries.sessionstorage.impl.di.SessionStorageModule.provideMatrixDatabase(SessionStorageModule.kt:40)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.libraries.sessionstorage.impl.di.SessionStorageModule_ProvideMatrixDatabaseFactory$Companion.provideMatrixDatabase(SessionStorageModule_ProvideMatrixDatabaseFactory.kt:32)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.libraries.sessionstorage.impl.di.SessionStorageModule_ProvideMatrixDatabaseFactory.get(SessionStorageModule_ProvideMatrixDatabaseFactory.kt:23)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.libraries.sessionstorage.impl.di.SessionStorageModule_ProvideMatrixDatabaseFactory.get(SessionStorageModule_ProvideMatrixDatabaseFactory.kt:20)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.libraries.sessionstorage.impl.DatabaseSessionStore_Factory.get(DatabaseSessionStore_Factory.kt:21)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.libraries.sessionstorage.impl.DatabaseSessionStore_Factory.get(DatabaseSessionStore_Factory.kt:18)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.libraries.matrix.impl.auth.RustMatrixAuthenticationService_Factory.get(RustMatrixAuthenticationService_Factory.kt:28)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.libraries.matrix.impl.auth.RustMatrixAuthenticationService_Factory.get(RustMatrixAuthenticationService_Factory.kt:21)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.appnav.di.MatrixClientsHolder_Factory.get(MatrixClientsHolder_Factory.java:30)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.appnav.di.MatrixClientsHolder_Factory.get(MatrixClientsHolder_Factory.java:11)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.x.di.DaggerAppComponent$AppComponentImpl.matrixClientsHolder(DaggerAppComponent.java:626)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at io.element.android.x.MainActivity.onCreate(MainActivity.kt:40)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.app.Activity.performCreate(Activity.java:6237)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1107)
2023-04-03 21:25:08.781  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2369)
2023-04-03 21:25:08.782  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	... 9 more
2023-04-03 21:25:08.782  4953-4953  VectorUncaughtException io.element.android.x.debug           E  Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
2023-04-03 21:25:08.782  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at libcore.io.Posix.open(Native Method)
2023-04-03 21:25:08.782  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at libcore.io.BlockGuardOs.open(BlockGuardOs.java:186)
2023-04-03 21:25:08.782  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	at libcore.io.IoBridge.open(IoBridge.java:438)
2023-04-03 21:25:08.782  4953-4953  VectorUncaughtException io.element.android.x.debug           E  	... 33 more
```
</details>

`./gradlew check`, `./gradlew ktlintCheck --continue`, and `./gradlew test` seem to pass. `./gradlew ktlintFormat` also runs without modifying anything.

### CODE REVIEW

File: features/login/impl/src/main/kotlin/io/element/android/features/login/impl/root/LoginRootView.kt
Comment (jmartinesp) [inline features/login/impl/src/main/kotlin/io/element/android/features/login/impl/root/LoginRootView.kt:370]: Could you move this to `io.element.android.libraries.designsystem.theme.components.TextField.kt` in the `:libraries:designsystem` module?

**Review by jmartinesp — CHANGES_REQUESTED** [review]
Thanks for the PR! This was something that was bothering me for a while but I never got the time to fix it.

The code looks good in general, and I got it to work (more or less) with a couple of password managers: there are some issues, but I believe these are just Compose issues and there's not much we can do about it.

However, there is a change I think we'd need to do before merging the code.

**Comment by jmartinesp** [comment]
Also, don't worry about the APK build job failure. We just need to rebase or merge develop once #292 is merged.

File: features/login/impl/src/main/kotlin/io/element/android/features/login/impl/root/LoginRootView.kt
Comment (hcsch) [reply features/login/impl/src/main/kotlin/io/element/android/features/login/impl/root/LoginRootView.kt]: Did the move in 61a1d219a2a63d538eeb62d8540fe424eec8eac2 :)

**Comment by hcsch** [comment]
Out of curiosity, what issue did you encounter with the autofilling that you'd attribute to Compose?

**Comment by jmartinesp** [comment]
> Out of curiosity, what issues did you encounter with the autofilling that you'd attribute to Compose?

Several ones:

- Autofill with 1Password didn't work for me at all when its UI should be displayed over the keyboard, but it worked when I set it to display a drowdown menu instead, although only when selecting the password field.
- With Enpass I didn't see the issue because it always displays the dropdown, but the position of the menu in the screen wasn't accurate some times.
- No support for saving data in autofill after a succesful login. I believe this issue has been around in Compose for a while, I remember seeing the same almost a year ago.

**Review by jmartinesp — APPROVED** [review]
LGTM, although we still need to wait for #292 to be merged so the CI will pass.

**Comment by hcsch** [comment]
I admittedly had only tested with KeePassDX, but that worked pretty well for me on both TextFields. To some degree, I've also experienced the autofill suggestion not showing above the keyboard in other apps, and only showing after unfocusing and refocusing the text field a couple of times. Nothing I'd know to do about that

The popup not showing up in the right place might be due to the TextField bounds not being of the correct kind / being somewhat inaccurate. In the `onGloballyPositioned` callback the code sets the bounds of the `AutofillNode` to those of the TextField inside the window, but the docs of `onGloballyPositioned` shown in Android Studio mention that a callback might not always be made if the position changes only relative to the screen, but not the window (the online docs https://developer.android.com/reference/kotlin/androidx/compose/ui/layout/LayoutCoordinates don't mention this for some reason, see quote from AS below). I'll see if better bounds can be obtained otherwise

`onGloballyPositioned` docs from Android Studio:
> Invoke onGloballyPositioned with the LayoutCoordinates of the element when the global position of the content may have changed. Note that it will be called after a composition when the coordinates are finalized.
This callback will be invoked at least once when the LayoutCoordinates are available, and every time the element's position changes within the window. However, it is not guaranteed to be invoked every time the position relative to the screen of the modified element changes. For example, the system may move the contents inside a window around without firing a callback. If you are using the LayoutCoordinates to calculate position on the screen, and not just inside the window, you may not receive a callback.

**Comment by hcsch** [comment]
Hmm... I couldn't really find any better callback for screen global bounds or similar after a cursory search. Not sure there is a better way to get and set the bounds. This is probably something for the Compose library to fix, so I'm gonna leave this code as-is

**Comment by jmartinesp** [comment]
> Hmm... I couldn't really find any better callback for screen global bounds or similar after a cursory search. Not sure there is a better way to get and set the bounds. This is probably something for the Compose library to fix, so I'm gonna leave this code as-is

Yeah, I also tried to improve it but I found no reliable way to do it, no worries.

Also, #292 was merged yesterday, so if you rebase or merge develop the CI should now pass.

**Pull request merged**
